// scopeField/ScopeFieldBase/ScopeFieldBase.js
import { Component, Children, cloneElement } from 'react';
import { PropTypes } from 'prop-types';

export default class ScopedField extends Component {
  componentDidUpdate(oldProps) {
    const { scope, resetFieldValue } = this.props;
    if (scope !== oldProps.scope) {
      resetFieldValue();
    }
  }

  render() {
    const {
      children,
      scope,
      scopeName,
      options: {
        allowEmptyScope,
        missingScopePlaceholderText
      } = {} } = this.props;
    const child = Children.only(children);
    const childProps = { ...child.props };

    let scopeRequiredButEmpty = !scope && !allowEmptyScope;
    if (scopeRequiredButEmpty) {
      childProps.disabled = true;
      childProps.placeholder = missingScopePlaceholderText || `Choose ${scopeName} first`;
    }

    return cloneElement(child, { ...childProps, key: scope });
  }
}

ScopedField.propTypes = {
  children: PropTypes.node.isRequired,
  scope: PropTypes.string,
  scopeName: PropTypes.string.isRequired,
  resetFieldValue: PropTypes.func.isRequired, // provided by wrapper
  onReset: PropTypes.func, // provided by user
  options: PropTypes.shape({
    allowEmptyScope: PropTypes.bool,
    missingScopePlaceholderText: PropTypes.string,
  })
};
